system-helper/flatpak-system-helper.c

```diff
@@ -237,16 +322,46 @@ handle_deploy (FlatpakSystemHelper   *object,
   gboolean local_pull;
   gboolean reinstall;
   g_autofree char *url = NULL;
+  g_autoptr(OngoingPull) ongoing_pull = NULL;
+  g_autofree gchar *src_dir = NULL;
+  gint dir_fd = -1;
 
   g_debug ("Deploy %s %u %s %s %s", arg_repo_path, arg_flags, arg_ref, arg_origin, arg_installation);
-
-  system = dir_get_system (arg_installation, get_sender_pid (invocation), &error);
+  system = dir_get_system ("", get_sender_pid (invocation), &error);
   if (system == NULL)
     {
       g_dbus_method_invocation_return_gerror (invocation, error);
       return TRUE;
     }
 
+  src_dir = g_path_get_dirname (arg_repo_path);
+  ongoing_pull = take_ongoing_pull_by_dir (src_dir);
+  if (ongoing_pull != NULL)
+    {
+      g_autoptr(GError) local_error = NULL;
+
+      if (shutdown (ongoing_pull->client_socket, SHUT_RDWR) == -1)
+        g_warning ("Failed to shutdown client socket, killing backend writer process anyway");
+
+      /* Killing the fuse backend will guarantee that all access to write operations are revoked. */
+      g_subprocess_force_exit (ongoing_pull->revokefs_backend);
```

Comment:
`g_subprocess_force_exit` doesn't wait for it to die. You also need `g_subprocess_wait()`.

Reply:
I think we discussed this somewhere (can't find where). It will get stuck on `g_subprocess_wait`. The backend is not a fork() process. Hence, we need to kill it in order to terminate it.

Reply:
Ah, I am wrong here. 
Shutting down the socket exits the --backend process. Apologies for creating noise.